internal/premis/premis.go

```diff
@@ -0,0 +1,375 @@
+	linkAgentIdentifierValueEl.CreateText(event.AgentIdValue)
+}
+
+func AddPREMISAgentElementIfNeeded(PREMISEl *etree.Element, agent PREMISAgent) {
+	if CheckIfPREMISAgentElementExists(PREMISEl, agent) {
+		return
+	}
+
+	// Add agent element.
+	agentEl := PREMISEl.CreateElement("premis:agent")
+
+	// Add agent identifier elements.
+	agentIdentifierEl := agentEl.CreateElement("premis:agentIdentifier")
+
+	agentIdentifierTypeEl := agentIdentifierEl.CreateElement("premis:agentIdentifierType")
+	agentIdentifierTypeEl.CreateText(agent.IdType)
+	agentIdentifierTypeEl.CreateAttr("valueURI", "http://id.loc.gov/vocabulary/identifiers/local")
+
+	agentIdentifierValueEl := agentIdentifierEl.CreateElement("premis:agentIdentifierValue")
+	agentIdentifierValueEl.CreateText(agent.IdValue)
+
+	// Add agent name and type.
+	agentNameEl := agentEl.CreateElement("premis:agentName")
+	agentNameEl.CreateText(agent.Name)
+
+	agentTypeEl := agentEl.CreateElement("premis:agentType")
+	agentTypeEl.CreateText(agent.Type)
+}
+
+func checkForDuplicatePREMISData(PREMISEl *etree.Element, elementTag string, paths map[string]string) bool {
+	// Cycle through agent elements so we can compare child data.
+	elements := PREMISEl.FindElements(fmt.Sprintf("//premis:%s", elementTag))
+
+	for _, element := range elements {
+		foundDifference := false
+
+		// Check child fields to see if they all match.
+		for path, value := range paths {
+			childEl := element.FindElement(fmt.Sprintf(".//%s[text()='%s']", path, value))
+
+			if childEl == nil {
+				foundDifference = true
+			}
+		}
+
+		if !foundDifference {
+			// Found match.
+			return true
+		}
+	}
+
+	// Found no match.
+	return false
+}
+
+func CheckIfPREMISObjectElementExists(PREMISEl *etree.Element, object PREMISObject) bool {
+	// Define xpath paths to child elements and values to match on.
+	paths := make(map[string]string)
+
+	paths["premis:object/premis:originalName"] = object.OriginalName
+
+	return checkForDuplicatePREMISData(PREMISEl, "object", paths)
+}
+
+func CheckIfPREMISAgentElementExists(PREMISEl *etree.Element, agent PREMISAgent) bool {
+	// Define xpath paths to child elements and values to match on.
+	paths := make(map[string]string)
+
+	paths["premis:agentType"] = agent.Type
+	paths["premis:agentName"] = agent.Name
+	paths["premis:agentIdentifier/premis:agentIdentifierType"] = agent.IdType
+	paths["premis:agentIdentifier/premis:agentIdentifierValue"] = agent.IdValue
+
+	return checkForDuplicatePREMISData(PREMISEl, "agent", paths)
+}
+
+func FilesWithinDirectory(contentPath string) ([]string, error) {
+	var subpaths []string
+
+	subpathStart := utf8.RuneCountInString(contentPath) + 1
+
+	err := filepath.WalkDir(contentPath, func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		subpaths = append(subpaths, string([]rune(p)[subpathStart:]))
```

Comment:
I find this solution a little hard to understand. Could we use https://pkg.go.dev/path/filepath#Rel?

Reply:
Fixed!